### PR TITLE
Empty strings in DIDSet transaction

### PR DESCRIPTION
**Pull request by ckeshava**
## High Level Overview of Change
Fix https://github.com/XRPLF/xrpl-py/issues/716

### Context of Change

Fields within a DID object can only be deleted by setting an empty string. However, the regex in the `DIDSet` model do not account for empty strings.

### Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### Did you update CHANGELOG.md?

- [x] Yes
- [ ] No, this change does not impact library users

## Test Plan

Unit tests have been added to this effect

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [[Unreleased]]
 - Add support for the DeliverMax field in Payment transactions
+- Allow empty strings for the purpose of removing fields in DIDSet transaction
```
Comment (mvadari) [inline CHANGELOG.md:10]: This should be under a `### Fixed` heading

File: xrpl/models/transactions/did_set.py [hunk 2]
```diff
@@ -37,6 +39,14 @@ def _get_errors(self: DIDSet) -> Dict[str, str]:
             # Can return here because there are no fields to process
             return errors
 
+        if self.did_document == "" and self.data == "" and self.uri == "":
+            errors["did_set"] = (
+                "At least one of DID fields `did_document`, `data`, and `uri` "
```
Comment (mvadari) [inline xrpl/models/transactions/did_set.py:44]:
```suggestion
                "At least one of the fields `did_document`, `data`, and `uri` "
```

File: xrpl/models/transactions/did_set.py [hunk 3]
```diff
@@ -12,7 +12,9 @@
 from xrpl.models.transactions.types import TransactionType
 from xrpl.models.utils import require_kwargs_on_init
 
-HEX_REGEX: Final[Pattern[str]] = re.compile("[a-fA-F0-9]+")
+# DID LedgerObject fields are removed with an empty string as input.
+# Hence, empty strings need to be supported in the regex
```
Comment (mvadari) [inline xrpl/models/transactions/did_set.py:16]: Code comments aren't useful if they're only relevant to the PR. This comment won't make any sense later on when looking at the code.

Comment (ckeshava) [reply xrpl/models/transactions/did_set.py:16]: I can remove this comment. But I believe it's useful.

It tells future developers why the regex shouldn't support empty strings.

Comment (ckeshava) [reply xrpl/models/transactions/did_set.py:16]: "... regex should support empty strings"

Comment (mvadari) [reply xrpl/models/transactions/did_set.py:16]: I think that should be discussed in the docs for each of the parameters, which should be copied from the docs here: https://xrpl.org/docs/references/protocol/transactions/types/didset

File: CHANGELOG.md (hunk 1 above)
Comment (ckeshava) [reply CHANGELOG.md:14]: fixed in [d0d5d75](https://github.com/XRPLF/xrpl-py/pull/717/commits/d0d5d753e528f5dacbd99e4bf04373d2ba6cb602)

File: xrpl/models/transactions/did_set.py (hunk 3 above)
Comment (ckeshava) [reply xrpl/models/transactions/did_set.py]: I agree. I've updated it in [e96a8ce](https://github.com/XRPLF/xrpl-py/pull/717/commits/e96a8cec419b1473086bfa76b0d996bdbb42b038)

File: tests/unit/models/transactions/test_did_set.py [hunk 4]
```diff
@@ -62,3 +62,46 @@ def test_empty(self):
                 error.exception.args[0],
                 "{'did_set': 'Must have one of `did_document`, `data`, and `uri`.'}",
             )
+
+    def test_all_empty_fields(self):
+        with self.assertRaises(XRPLModelException):
+            DIDSet(account=_ACCOUNT, data="", did_document="", uri="")
+
+    def test_delete_multiple_fields(self):
```
Comment (mvadari) [inline tests/unit/models/transactions/test_did_set.py:70]: super nit:
```suggestion
    def test_remove_multiple_fields(self):
```
I was a bit confused at first because I thought this might be a `DIDDelete` test.

File: tests/unit/models/transactions/test_did_set.py [hunk 5]
```diff
@@ -62,3 +62,46 @@ def test_empty(self):
...
+        # create a valid DID object
+        tx = DIDSet(
+            account=_ACCOUNT,
+            did_document=_VALID_FIELD,
+            uri=_VALID_FIELD,
+            data=_VALID_FIELD,
+        )
+        self.assertTrue(tx.is_valid())
```
Comment (mvadari) [inline tests/unit/models/transactions/test_did_set.py:78]: This is unnecessary because you're not actually sending the transaction.

File: tests/unit/models/transactions/test_did_set.py [hunk 6]
```diff
@@ -62,3 +62,46 @@ def test_empty(self):
...
+        )
+
+        self.assertTrue(tx.is_valid())
+
+        # delete did_document field
+
+        # Note: It is invalid to render a DID LedgerObject empty by removing all the
+        # fields. However, that condition can only be detected by reading the
+        # LedgerEntry, which is out-of-scope for the library. rippled generated a
+        # `tecEMPTY_DID` error code in this case
```
Comment (mvadari) [inline tests/unit/models/transactions/test_did_set.py:101]: This is also unnecessary since the transaction is never actually submitted.

File: xrpl/models/transactions/did_set.py [hunk 7]
```diff
@@ -21,8 +21,24 @@ class DIDSet(Transaction):
     """Represents a DIDSet transaction."""
 
     did_document: Optional[str] = None
+    """
+    The DID document associated with the DID.
+    You must include either Data, DIDDocument, or URI when you submit the DIDSet
+    transaction. If all three fields are missing, the transaction fails.
+
+    Note: To delete the Data, DIDDocument, or URI field from an existing DID ledger
+    entry, add the field as an empty string.
```
Comment (mvadari) [inline xrpl/models/transactions/did_set.py:30]:
```suggestion
    The DID document associated with the DID.
    To delete this field from an existing DID ledger entry, include it as an empty string.
```

File: xrpl/models/transactions/did_set.py [hunk 8]
```diff
@@ -21,8 +21,24 @@ class DIDSet(Transaction):
...
+    """
+
     data: Optional[str] = None
+    """
+    The public attestations of identity credentials associated with the DID.
```
Comment (mvadari) [inline xrpl/models/transactions/did_set.py:35]:
```suggestion
    The public attestations of identity credentials associated with the DID.
    To delete this field from an existing DID ledger entry, include it as an empty string.
```

File: xrpl/models/transactions/did_set.py [hunk 9]
```diff
@@ -21,8 +21,24 @@ class DIDSet(Transaction):
...
+    """
+
     uri: Optional[str] = None
+    """
+    The Universal Resource Identifier associated with the DID.
```
Comment (mvadari) [inline xrpl/models/transactions/did_set.py:40]:
```suggestion
    The Universal Resource Identifier associated with the DID.
    To delete this field from an existing DID ledger entry, include it as an empty string.
```

File: tests/unit/models/transactions/test_did_set.py (hunk 5 above)
Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py:78]: are you proposing that I remove this entire unit test? Or just the `self.assertTrue(...)` statement?

File: tests/unit/models/transactions/test_did_set.py (hunk 6 above)
Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py:101]: Yeah, but I wanted to exercise the `_get_errors` function in the `DIDSet` model. Otherwise, we will probably need to do the same with an integration test

**Review by khancode — CHANGES_REQUESTED** [review]
Add integration test for submitting `DIDSet` with empty string.

File: tests/unit/models/transactions/test_did_set.py (hunk 6 above)
Comment (mvadari) [reply tests/unit/models/transactions/test_did_set.py:101]: `_get_errors` runs on model creation, not on transaction submission. I was referring to the comment, not the transaction, since it's not actually submitting the transaction.

File: tests/unit/models/transactions/test_did_set.py [hunk 10]
```diff
@@ -62,3 +62,46 @@ def test_empty(self):
                 error.exception.args[0],
                 "{'did_set': 'Must have one of `did_document`, `data`, and `uri`.'}",
             )
+
+    def test_all_empty_fields(self):
+        with self.assertRaises(XRPLModelException):
+            DIDSet(account=_ACCOUNT, data="", did_document="", uri="")
+
+    def test_remove_multiple_fields(self):
```
Comment (mvadari) [inline tests/unit/models/transactions/test_did_set.py:70]: IMO each transaction should be a separate test. Easier to run.

Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py:70]: I'd like to remove consecutive fields in the DID object, in order to demonstrate that users cannot leave behind an empty DID object.

I need to perform consecutive removals on the same object, it cannot be demonstrated with a new DID ledger object

Comment (mvadari) [reply tests/unit/models/transactions/test_did_set.py:70]: That's not a library test, that's a rippled test. And even if it were a library test, it should happen in an integration test, because it won't happen in a unit test (the unit test doesn't submit transactions and therefore doesn't create/modify/delete ledger objects). All you're doing here is creating the models, not submitting anything.

File: tests/unit/models/transactions/test_did_set.py (hunk 6 above)
Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py]: I've removed the comment and the test in [ed1c57c](https://github.com/XRPLF/xrpl-py/pull/717/commits/ed1c57c3e995b02cceb4b70a5e587dc7cfae3c52)

File: tests/unit/models/transactions/test_did_set.py (hunk 10 above)
Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py]: okay, I've created an integration test for the same

**Comment by ckeshava** [comment]
I have reverted back to an older commit. I removed the integration test and restored the unit test. Let me know if this looks ok

File: tests/unit/models/transactions/test_did_set.py [hunk 11]
```diff
@@ -62,3 +62,25 @@ def test_empty(self):
                 error.exception.args[0],
                 "{'did_set': 'Must have one of `did_document`, `data`, and `uri`.'}",
             )
+
+    def test_all_empty_fields(self):
+        with self.assertRaises(XRPLModelException):
+            DIDSet(account=_ACCOUNT, data="", did_document="", uri="")
+
+    def test_delete_multiple_fields(self):
```
Comment (khancode) [inline tests/unit/models/transactions/test_did_set.py:70]: I would keep the wording consistent with "empty fields" and not use "delete" since this is a unit test.

File: tests/unit/models/transactions/test_did_set.py [hunk 12]
```diff
@@ -62,3 +62,25 @@ def test_empty(self):
...
+        tx = DIDSet(
+            account=_ACCOUNT,
+            did_document=_VALID_FIELD,
+            uri=_VALID_FIELD,
+            data=_VALID_FIELD,
+        )
+        self.assertTrue(tx.is_valid())
+
+        # remove the data field from the above DID object
+        tx = DIDSet(
```
Comment (mvadari) [inline tests/unit/models/transactions/test_did_set.py:81]: Do you still want a transaction for each field?

File: tests/unit/models/transactions/test_did_set.py [hunk 13]
```diff
@@ -62,3 +62,25 @@ def test_empty(self):
...
                 "{'did_set': 'Must have one of `did_document`, `data`, and `uri`.'}",
             )
+
+    def test_all_empty_fields(self):
+        with self.assertRaises(XRPLModelException):
+            DIDSet(account=_ACCOUNT, data="", did_document="", uri="")
+
+    def test_delete_multiple_fields(self):
+        # create a valid DID object
+        tx = DIDSet(
```
Comment (mvadari) [inline tests/unit/models/transactions/test_did_set.py:72]: All of these transactions should be in separate tests.

File: tests/unit/models/transactions/test_did_set.py (hunk 11 above)
Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py]: updated in [1b7400a](https://github.com/XRPLF/xrpl-py/pull/717/commits/1b7400a0f12a732cfaf79f521cf69c3199b33c4f)

File: tests/unit/models/transactions/test_did_set.py (hunk 13 above)
Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py:72]: there are only two transactions -- create a `DID` object, and then remove the `data` field from the said object.

I cannot decouple them into separate unit tests, I need a `DID` ledger object to demonstrate the removal of the `data` field.

File: tests/unit/models/transactions/test_did_set.py (hunk 12 above)
Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py:81]: I'm only removing the `data` field. I'm not using any other transaction

File: tests/unit/models/transactions/test_did_set.py (hunk 13 above)
Comment (mvadari) [reply tests/unit/models/transactions/test_did_set.py:72]: It's a unit test, the ledger object is never created in the first place.

File: tests/unit/models/transactions/test_did_set.py (hunk 12 above)
Comment (mvadari) [reply tests/unit/models/transactions/test_did_set.py:81]: You originally also removed the other fields.

File: tests/unit/models/transactions/test_did_set.py (hunk 11 above)
Comment (khancode) [reply tests/unit/models/transactions/test_did_set.py]: I would use "empty fields" rather than "remove" since this is a unit test.

File: tests/unit/models/transactions/test_did_set.py (hunk 13 above)
Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py:72]: okay, that's true.

should I remove this unit test altogether? it appears that rippled would have tested this piece of functionality.
I don't see other ways of testing this through a unit test

File: tests/unit/models/transactions/test_did_set.py (hunk 12 above)
Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py:81]: yes, that's correct. but now, I'm only empty-ing the `data` field

File: tests/unit/models/transactions/test_did_set.py (hunk 13 above)
Comment (mvadari) [reply tests/unit/models/transactions/test_did_set.py:72]: Testing that one empty field works is still important.

File: tests/unit/models/transactions/test_did_set.py (hunk 12 above)
Comment (mvadari) [reply tests/unit/models/transactions/test_did_set.py:81]: Yes, should there be tests for emptying each field?

Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py:81]: can you tell me the name of the unit test that you are objecting to? 

`test_empty_data_field` -- a unit test that empties only the `data` field. all the other fields contain valid values.

`test_create_did_object_all_empty_fields` --  a unit test that validates the initial input for the `DIDSet` transaction. This verifies that all the input fields are not empty.


You are correct that rippled has already tested this functionality. Which one of these warrants a change?

File: tests/unit/models/transactions/test_did_set.py (hunk 13 above)
Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py:72]: okay. you mean like creating a DID object with one empty field?

```
def test_single_empty_field(self):
        # create a valid DID object
        tx = DIDSet(
            account=_ACCOUNT,
            did_document=_VALID_FIELD,
            uri="", # URI field is empty
            data=_VALID_FIELD,
        )
        self.assertTrue(tx.is_valid())
```

Comment (mvadari) [reply tests/unit/models/transactions/test_did_set.py:72]: Yes. Or without including the `did_document` and `data` fields.

File: tests/unit/models/transactions/test_did_set.py (hunk 5 above)
Comment (mvadari) [reply tests/unit/models/transactions/test_did_set.py:78]: The transaction.

File: tests/unit/models/transactions/test_did_set.py (hunk 13 above)
Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py:72]: I gave this input to one of the integration tests:
```
tx = DIDSet(account=WALLET.address, uri="")
```

This fails with an error -- we cannot input an empty URI field, while other fields are also non-existent.
```
======================================================================
FAIL: test_all_fields (tests.integration.transactions.test_did_set.TestDIDSet.test_all_fields) (version='async', client='json')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/ckeshavabs/xrpl-py/tests/integration/it_utils.py", line 290, in modified_test
    asyncio.run(
  File "/Users/ckeshavabs/.pyenv/versions/3.11.6/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/Users/ckeshavabs/.pyenv/versions/3.11.6/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ckeshavabs/.pyenv/versions/3.11.6/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/Users/ckeshavabs/xrpl-py/tests/integration/it_utils.py", line 281, in _run_async_test
    raise e
  File "/Users/ckeshavabs/xrpl-py/tests/integration/it_utils.py", line 277, in _run_async_test
    await test_function(self, client)
  File "/Users/ckeshavabs/xrpl-py/tests/integration/transactions/test_did_set.py", line 22, in test_all_fields
    self.assertEqual(response.result["engine_result"], "tesSUCCESS")
AssertionError: 'tecEMPTY_DID' != 'tesSUCCESS'
- tecEMPTY_DID
+ tesSUCCESS
```

Comment (mvadari) [reply tests/unit/models/transactions/test_did_set.py:72]: You're running it as an integration test, not a unit test. I don't need this to be tested as an integration test, I know that's going to work. That's what rippled unit tests are for. I just want to make sure that xrpl-py accepts it as a valid transaction, without knowing anything about the ledger state.

Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py:72]: okay. I've updated this in 10f7f38

I am slightly opposed to this type of tests. I'd prefer to have unit tests that are in consonance with the integration tests too, that reduces ambiguity for future developers

Comment (mvadari) [reply tests/unit/models/transactions/test_did_set.py:72]: > that reduces ambiguity for future developers

What do you mean by this?

The way I think about it is that unit tests are checking xrpl-py Python code and integration tests are ensuring that xrpl-py Python code matches rippled code. So for checking xrpl-py error code, you only want to write unit tests.

Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py:72]: that is true.

When looking at the unit tests, I'm led to believe that a DID object can be created with empty fields. But that is not the case with a real rippled instance.

I feel this contradiction could have been avoided, but it's not a strong objection

Comment (mvadari) [reply tests/unit/models/transactions/test_did_set.py:68]: I would assume anyone advanced enough to be looking at unit tests to understand the system is probably well-versed enough in the XRPL that they're not going to fall into that kind of trap - or will at least know that unit tests aren't being submitted to the ledger.

Comment (ckeshava) [reply tests/unit/models/transactions/test_did_set.py:68]: fair point

**Review by mvadari — APPROVED** [review]

**Review by khancode — APPROVED** [review]
